numpyro/distributions/constraints.py

```diff
@@ -242,31 +241,32 @@ def __call__(
             event_dim = self._event_dim
         return _Dependent(is_discrete=is_discrete, event_dim=event_dim)
 
-    def __eq__(self, other: ConstraintT) -> bool:
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, _Dependent):
+            return False
         return (
-            type(self) is type(other)
-            and self._is_discrete == other._is_discrete
+            self._is_discrete == other._is_discrete
             and self._event_dim == other._event_dim
         )
 
     def tree_flatten(self):
         return (), (
             (),
-            dict(_is_discrete=self._is_discrete, _event_dim=self._event_dim),
+            dict(_is_discrete=self._is_discrete, _event_dim=self.event_dim),
         )
 
 
-class dependent_property(property, _Dependent):
+class dependent_property(property, _Dependent[NumLikeT]):
     # XXX: this should not need to be pytree-able since it simply wraps a method
     # and thus is automatically present once the method's object is created
     def __init__(
         self, fn=None, *, is_discrete=NotImplemented, event_dim=NotImplemented
     ):
         super().__init__(fn)
         self._is_discrete = is_discrete
-        self._event_dim = event_dim
+        self.event_dim = event_dim
 
-    def __call__(self, x: ArrayLike) -> ArrayLike:
+    def __call__(self, x: NumLikeT) -> ArrayLike:
```

Comment:
x here could be a function, and the return value could be a function. Maybe ignore type here
```
def __call__(self, x):
```